polygeist/lib/Dialect/Polygeist/Utils/TransformUtils.cpp

```diff
@@ -20,6 +22,97 @@ using namespace mlir;
 // Utilities functions
 //===----------------------------------------------------------------------===//
 
+bool mlir::isLinkonceODR(FunctionOpInterface func) {
+  constexpr StringRef linkageAttrName = "llvm.linkage";
+  if (!func->hasAttr(linkageAttrName))
+    return false;
+  auto attr = cast<LLVM::LinkageAttr>(func->getAttr(linkageAttrName));
+  return attr.getLinkage() == LLVM::Linkage::LinkonceODR;
+}
+
+bool mlir::isTailCall(CallOpInterface call) {
+  if (!call->getBlock()->hasNoSuccessors())
+    return false;
+  return isRegionReturnLike(call->getNextNode());
+}
+
+/// Populate \p funcMaxDepthMap with the maximum depth from a GPU kernel for \p
+/// func and its callers.
+static void
+getMaxDepthFromGPUKernel(FunctionOpInterface func,
+                         DenseMap<FunctionOpInterface, int> &funcMaxDepthMap) {
+  assert(!funcMaxDepthMap.contains(func) &&
+         "Expecting maximum depth of func is not already calculated");
+
+  // A function that does not reside in a GPU module cannot be called from a GPU
+  // kernel.
+  if (!func->getParentOfType<gpu::GPUModuleOp>()) {
+    funcMaxDepthMap[func] = -1;
+    return;
+  }
```

Comment:
I was thinking that we might want to track the depth in "both directions", therefore the host function that contain a `gpu.launch` operation should have depth -1, and its caller should have depth -2, etc... 

I realize we currently do not have access to the host code. How about returning an Optional depth. It can return std::nullopt when the depth is actually unknown. In the future I think it will be easier to extend this code.

Reply:
done.